File: src/gp/gpnode.c
/*
 * Copyright (C) 2010 Telmo Menezes.
 * telmo@telmomenezes.com
 */


#include <stdlib.h>
#include <stdio.h>
#include "gpnode.h"


gpnode* create_gpnode(gpnode_type nodetype,
                        gpnode_fun fun,
                        gpval val,
                        unsigned int var,
                        gpnode* parent)
{
    gpnode* node = (gpnode *)malloc(sizeof(gpnode));
    node->type = nodetype;
    node->parent = parent;

    if (nodetype == FUN) {
        node->fun = fun;
        node->arity = fun_arity(fun);
        node->condpos = fun_condpos(fun);
    }
    else if (nodetype == VAR) {
        node->var = var;
        node->arity = 0;
        node->condpos = -1;
    }
    else {
        node->val = val;
        node->arity = 0;
        node->condpos = -1;
    }

    node->stoppos = node->arity;

    return node;
}


void destroy_gpnode(gpnode* node)
{
    free(node);
}


int fun_condpos(gpnode_fun fun)
{
    switch(fun) {
        case ZER:
            return 1;
        case EQ:
        case GRT:
        case LRT:
            return 2;
        default:
            return -1;
    }
}


unsigned int fun_arity(gpnode_fun fun)
{
    switch(fun) {
        case EXP:
        case LOG:
        case SIN:
        case ABS:
            return 1;
        case SUM:
        case SUB:
        case MUL:
        case DIV:
            return 2;
        case ZER:
            return 3;
        case EQ:
        case GRT:
        case LRT:
            return 4;
        default:
            return 0;
    }
}


void print_gpnode(gpnode* node)
{
    if (node->type == VAL) {
        printf("%f", node->val);
        return;
    }

    if (node->type == VAR) {
        printf("$%d", node->var);
        return;
    }

    if (node->type != FUN) {
        printf("???");
        return;
    }

    switch(node->fun) {
        case SUM:
            printf("+");
            return;
        case SUB:
            printf("-");
            return;
        case MUL:
            printf("*");
            return;
        case DIV:
            printf("/");
            return;
        case ZER:
            printf("ZER");
            return;
        case EQ:
            printf("==");
            return;
        case GRT:
            printf(">");
            return;
        case LRT:
            printf("<");
            return;
        case EXP:
            printf("EXP");
            return;
        case LOG:
            printf("LOG");
            return;
        case SIN:
            printf("SIN");
            return;
        case ABS:
            printf("ABS");
            return;
        default:
            printf("F??");
            return;
    }
}

File: src/com/telmomenezes/synthetic/gp/GPNode.java
package com.telmomenezes.synthetic.gp;

import java.io.IOException;
import java.io.OutputStreamWriter;

/**
 * Genetic program node.
 * 
 * @author Telmo Menezes (telmo@telmomenezes.com)
 */
public class GPNode {

    public GPNodeType type;
    public double val;
    public int var;
    public int fun;
    public int arity;
    public GPNode[] params;
    public double curval;
    protected GPNode parent;

    public int curpos;
    public int stoppos;
    public int condpos;

    public int branching;
    public GPNodeDynStatus dynStatus;
    
    public int evals;
    public int lastEval;
    
    private Prog tree;

    public GPNode(Prog tree) {
    	this.tree = tree;
        params = new GPNode[4];
    }

    public void initVal(double val, GPNode parent) {
        type = GPNodeType.VAL;
        this.parent = parent;
        this.val = val;
        arity = 0;
        condpos = -1;
        stoppos = 0;
        dynStatus = GPNodeDynStatus.UNUSED;
    }

    public void initVar(int var, GPNode parent) {
        type = GPNodeType.VAR;
        this.parent = parent;
        this.var = var;
        arity = 0;
        condpos = -1;
        stoppos = 0;
        dynStatus = GPNodeDynStatus.UNUSED;
    }

    public void initFun(int fun, GPNode parent) {
        type = GPNodeType.FUN;
        this.parent = parent;
        this.fun = fun;
        arity = funArity(fun);
        condpos = funCondPos(fun);
        stoppos = arity;
        dynStatus = GPNodeDynStatus.UNUSED;
    }

    private int funCondPos(int fun) {
        switch (fun) {
        case GPFun.ZER:
        case GPFun.GRP:
            return 1;
        case GPFun.EQ:
        case GPFun.GRT:
        case GPFun.LRT:
        case GPFun.MOD:
            return 2;
        default:
            return -1;
        }
    }

    private int funArity(int fun) {
        switch (fun) {
        case GPFun.EXP:
        case GPFun.LOG:
        case GPFun.ABS:
            return 1;
        case GPFun.SUM:
        case GPFun.SUB:
        case GPFun.MUL:
        case GPFun.DIV:
        case GPFun.MIN:
        case GPFun.MAX:
            return 2;
        case GPFun.ZER:
        case GPFun.GRP:
            return 3;
        case GPFun.EQ:
        case GPFun.GRT:
        case GPFun.LRT:
        case GPFun.MOD:
            return 4;
        // this should not happen
        default:
            return 0;
        }
    }

    public void write(OutputStreamWriter out, boolean evalStats)
            throws IOException {
        if (type == GPNodeType.VAL) {
            out.write("" + val);
        }
        else if (type == GPNodeType.VAR) {
            out.write("$" + tree.getVariableNames().get(var));
        }
        else if (type == GPNodeType.FUN) {
            switch (fun) {
            case GPFun.SUM:
                out.write("+");
                break;
            case GPFun.SUB:
                out.write("-");
                break;
            case GPFun.MUL:
                out.write("*");
                break;
            case GPFun.DIV:
                out.write("/");
                break;
            case GPFun.ZER:
                out.write("ZER");
                break;
            case GPFun.EQ:
                out.write("==");
                break;
            case GPFun.GRT:
                out.write(">");
                break;
            case GPFun.LRT:
                out.write("<");
                break;
            case GPFun.EXP:
                out.write("EXP");
                break;
            case GPFun.LOG:
                out.write("LOG");
                break;
            case GPFun.ABS:
                out.write("ABS");
                break;
            case GPFun.MIN:
                out.write("MIN");
                break;
            case GPFun.MAX:
                out.write("MAX");
                break;
            case GPFun.MOD:
                out.write("%");
                break;
            case GPFun.GRP:
                out.write("GRP");
                break;
            default:
                out.write("F??");
                break;
            }
        }
        else {
            out.write("???");
        }
        
        if (evalStats) {
            out.write(" " + evals + " " + lastEval);
        }
    }
}